# 1.Programming Assignment 1: Least Sqaures

ANLT_201_01-Linear Alegbra for Data Sci Hybrid/Blended-80942

Group 4 : Jiye Lyu and Harsh Rajeshkumar Shah

## 1.1Objective:

We are given data with various factors describing Quarterbacks' performance. It also includes rating points. We have to 
analyse the data and create Least Square model which predicts quarterback ratings using various factors. 

In [187]:
#Loading data
data <- read.csv('rating.csv')

In [228]:
#Finding dimensions
dim(data)

[1] 32 14

We have 32 rows and 14 columns

In [229]:
#Printing first few rows
head(data)

First  Last       Team Att Comp Pct.Comp Yds  Yds.per.Att TD Pct.TD Long Int
1 Philip Rivers     SD   478 312  65.3     4009 8.39        34 7.1    67   11 
2 Chad   Pennington MIA  476 321  67.4     3653 7.67        19 4.0    80    7 
3 Kurt   Warner     ARI  598 401  67.1     4583 7.66        30 5.0    79   14 
4 Drew   Brees      NO   635 413  65.0     5069 7.98        34 5.4    84   17 
5 Peyton Manning    IND  555 371  66.8     4002 7.21        27 4.9    75   12 
6 Aaron  Rodgers    GB   536 341  63.6     4038 7.53        28 5.2    71   13 
  Pct.Int Rating.Pts
1 2.3     105.5     
2 1.5      97.4     
3 2.3      96.9     
4 2.7      96.2     
5 2.2      95.0     
6 2.4      93.8

As it can be seen, the table includes players' name and the data that can be used to assess the performance of players. 

## 2.Creating Least Square Models

### 2.1Model A: Using Percentage of Completions(Pct.Comp)

In this model, we have to predict quarterback ratings using their percentage of completions. So here (Pct.Comp, Rating.Pts) are
points in a plane and least square model creates a line which is fitted to best approximate the points. It is done by drawing a line parallel to y-axis from the data points to the line that measures the error of each data point. The sum of all such errors should be minimum. 

In [201]:
x1 <- as.matrix(cbind(rep(1, nrow(data)), data$Pct.Comp))
head(x1)

[,1] [,2]
[1,] 1    65.3
[2,] 1    67.4
[3,] 1    67.1
[4,] 1    65.0
[5,] 1    66.8
[6,] 1    63.6

In [200]:
y1 <- as.matrix(data$Rating.Pts)
head(y1)

[,1] 
[1,] 105.5
[2,]  97.4
[3,]  96.9
[4,]  96.2
[5,]  95.0
[6,]  93.8

In [234]:
const1 = inv(t(x1) %*% x1) %*% t(x1) %*% y1
const1

[,1]      
[1,] -22.033064
[2,]   1.745626

The best line according to this model is given by the equation:

y = c0 + c1(x1) 

where c0 = -22.033064 and c1 = 1.745626

### 2.2Model B: Using Percentage of Completions(Pct.Comp) and Interceptions(Pct.Int)

In this model, we have to predict quarterback ratings using their percentage of completions and percentage of interceptions. So here (Pct.Comp, Rating.Pts) and (Pct.Int, Rating.Pts) are points in a plane and least square model creates a plane which is fitted to best approximate the points. It is done by drawing a line perpendicular to plane from the data points that measures the error of each data point. The sum of all such errors should be minimum. 

In [203]:
x2 <- as.matrix(cbind(rep(1, nrow(data)), data$Pct.Comp, data$Pct.Int ))
head(x2)

[,1] [,2] [,3]
[1,] 1    65.3 2.3 
[2,] 1    67.4 1.5 
[3,] 1    67.1 2.3 
[4,] 1    65.0 2.7 
[5,] 1    66.8 2.2 
[6,] 1    63.6 2.4

In [204]:
y2 <- as.matrix(data$Rating.Pts)
head(y2)

[,1] 
[1,] 105.5
[2,]  97.4
[3,]  96.9
[4,]  96.2
[5,]  95.0
[6,]  93.8

In [235]:
const2 = inv(t(x2) %*% x2) %*% t(x2) %*% y2
const2

[,1]     
[1,] -9.109043
[2,]  1.662173
[3,] -3.076368

The best plane according to this model is given by the equation:

y2 = c0 + c1(x1) + c2(x2) 

where c0 = -9.109043, c1 = 1.662173 and c2 = -3.076368

## 3.Findng Least Square Error

### 3.1For Model A

In [230]:
error1 <- norm(y1 - x1 %*% const1, '2')
cat("The least square error in Model A is ", error1)

The least square error in Model A is  29.58189

### 3.2For Model B

In [231]:
error2 <- norm(y2 - x2 %*% const2, '2')
cat("The least square error in Model B is ", error2)

The least square error in Model B is  26.58117

As the error gets reduced, we can say that using the percentage of interceptions additionally improves the accuracy of the model and can improve the predictions of quarterback ratings.

## 4.Prediction

In [215]:
test_pct_comp <- 60
test_pct_int <- 3

### 4.1Using Model A 

In [219]:
modelA_pred <- c(1, test_pct_comp) %*% const1
modelA_pred

[,1]    
[1,] 82.70451

In [224]:
cat("Using Model A, a quarterback with 60% completions will get ratings", modelA_pred)

Using Model A, a quarterback with 60% completions will get ratings 82.70451

### 4.2Using Model B

In [221]:
modelB_pred <- c(1, test_pct_comp, test_pct_int) %*% const2
modelB_pred

[,1]    
[1,] 81.39223

In [223]:
cat("Using Model B, a quarterback with 60% completions and 3% interceptions will get ratings", modelB_pred)

Using Model B, a quarterback with 60% completions and 3% interceptions will get ratings 81.39223